src/agent/agent.ts

```diff
@@ -567,67 +567,103 @@ export class Agent implements AgentData {
       input: toolUseBlock.input,
     }
 
-    yield new BeforeToolCallEvent({ agent: this, toolUse, tool })
+    // Yield BeforeToolCallEvent and capture potentially modified values
+    const beforeEvent = new BeforeToolCallEvent({ agent: this, toolUse, tool })
+    yield beforeEvent
 
-    if (!tool) {
+    // Use potentially modified tool and toolUse from hook
+    let actualTool = beforeEvent.tool
+    const actualToolUse = beforeEvent.toolUse
+
+    // If toolUse.name was modified, re-lookup the tool
+    if (actualToolUse.name !== toolUseBlock.name) {
+      actualTool = toolRegistry.find((t) => t.name === actualToolUse.name)
+    }
```

Comment:
Don't do this; if they want to override the tool, they have to set `tool`, not set the `toolUseBlock`.

Make the change, and ensure this is documented